### PR TITLE
GH-102973: add a dev container

### PR DESCRIPTION
**Pull request by brettcannon**
Defines a Dockerfile based on Fedora along with common tooling (e.g. various shells people may want to use). It also includes the tooling necessary to do a cross-build for WASI.

Co-authored-by: Dusty Phillips <dusty@phillips.codes>

* Issue: gh-102973

### CODE REVIEW

**Comment by frenzymadness** [comment]
Thanks @encukou for the ping. I like it. You might want to use `--setopt=tsflags=nodocs` for the `dnf` command to skip documentation packages (`--no-docs` should work as well) and then you might want to do `dnf -y clean all` to remove cached files from the resulting image.

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                    "*.h": "c"
+                },
+                "files.encoding": "utf8",
+                "files.eol": "\n",
+                "files.insertFinalNewline": true,
+                "files.trimTrailingWhitespace": true,
+                "python.analysis.diagnosticSeverityOverrides": {
+                    // Complains about shadowing the stdlib w/ the stdlib.
+                    "reportShadowedImports": "none",
+                    // Dodesn't like _frozen_importlib.
```
Comment (ronaldoussoren) [inline .devcontainer/devcontainer.json]:
```suggestion
                    // Doesn't like _frozen_importlib.
```

File: Tools/wasm/Setup.local.example [hunk 2]
```diff
@@ -5,6 +5,7 @@ audioop
 _bz2
 _crypt
 _decimal
+nis
```
Comment (ronaldoussoren) [inline Tools/wasm/Setup.local.example:8]: This change looks unrelated to the rest of the PR.

Comment (brettcannon) [reply Tools/wasm/Setup.local.example:8]: Tehnically, yes, but while testing this it led to https://github.com/python/cpython/issues/102851 which this would close.

**Comment by frenzymadness** [comment]
> Thanks @frenzymadness ! PTAL and let me know if I got everything you were suggesting.

Looks perfect to me now.

**Comment by encukou** [comment]
I'm not familiar with devcontainers, let me throw ideas around for consideration.

I don't understand the `onCreateCommand`. Is in any different from the `Dockerfile` install section? Installing everything at once could save some repo syncing.

You might want to install `ccache` to speed up builds, especially after switching branches. The default cache size (5G, AFAIK) might be too much.

You might want to install `python3-blurb` from RPM, if avoiding yet another tool (`pipx`) sounds interesting.

There's also a  `python3-sphinx` RPM. With it you could `make --directory Doc html` without `venv` to save time. If you want to add it, install with `dnf --setopt=install_weak_deps=False` to skip dependencies related to image manipulation.

**Comment by brettcannon** [comment]
> I don't understand the `onCreateCommand`. Is in any different from the `Dockerfile` install section? Installing everything at once could save some repo syncing.

It is, but I view the dev container details as specific to editing via GitHub Codespaces while with the Dockerfile I tried to take a less opinionated approach.



> You might want to install `ccache` to speed up builds, especially after switching branches. The default cache size (5G, AFAIK) might be too much.

👍



> You might want to install `python3-blurb` from RPM, if avoiding yet another tool (`pipx`) sounds interesting.

I'm up for it!



> There's also a `python3-sphinx` RPM. With it you could `make --directory Doc html` without `venv` to save time.

With GitHub Codespaces prebuilds the save is minimal since the assumption is everyone will start from the container already created and all of those commands already run (takes less then 45 seconds to get to a terminal prompt). But I'm up for trying it out!

**Comment by brettcannon** [comment]
Turns out `python3-sphinx` is the wrong Sphinx version and skips installing the theme and other things required to build the docs, so I don't think it's going to improve anything.

**Comment by brettcannon** [comment]
I decided to not go with `ccache` to start since it could lead to actual costs for people who might not be aware that `ccache` is running and upping the amount of storage their Codespace is using (granted, people get 15GB free every month, but still).

**Comment by brettcannon** [comment]
I decided to go ahead and merge this as we can easily update it if there's anything we want to tweak.

**Comment by encukou** [comment]
In case you want to revisit `ccache` (e.g. if juggling branches is a common use case), the cache size can be set with `echo 'max_size = 100M' > /etc/ccache.conf`
* I'm assuming the devcontainer script controls system-wide setup (`/etc/`): user-specific config should go to `$HOME/.config/ccache/ccache.conf` & Fedora defaults are compiled in
* A full build takes up ~27M of cache, two very different branches (3.12 and 3.6) fill ~40M.

> Turns out python3-sphinx is the wrong Sphinx version and skips installing the theme and other things required to build the docs, so I don't think it's going to improve anything.

Fair. I use it (one less venv → one less update command to forget to run, and with docs previews a slightly outdated version doesn't matter that much), but, it's not for everyone.
But in case it's useful: to install the theme, linkchecker, and other things, there's `dnf builddep python3-docs`

**Comment by brettcannon** [comment]
> In case you want to revisit `ccache` (e.g. if juggling branches is a common use case), the cache size can be set with `echo 'max_size = 100M' > /etc/ccache.conf`

Yep, I believe you can also use `ccache --max-size`, but since that decision could cost people money, I didn't want to make that decision for them.



> (one less venv → one less update command to forget to run, and with docs previews a slightly outdated version doesn't matter that much

True, although once we get prebuilds up and running that will be automatic after every commit to `main`.

**Comment by brettcannon** [comment]
Prebuilds have been turned on! https://github.com/python/cpython/actions/workflows/codespaces/create_codespaces_prebuilds

/cc @methane

File: .devcontainer/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+ENV CC=clang
+
+ENV WASI_SDK_VERSION=19
+ENV WASI_SDK_PATH=/opt/wasi-sdk
+
+ENV WASMTIME_HOME=/opt/wasmtime
+ENV WASMTIME_VERSION=7.0.0
+ENV WASMTIME_CPU_ARCH=x86_64
+
+RUN dnf -y --nodocs install git clang xz python3-blurb dnf-plugins-core && \
```
Comment (hroncok) [inline .devcontainer/Dockerfile:12]: I realize I am late to the party, but you might save space by installing git-core instead of git, which should avoid Perl (unless something else brings it in).
(`dnf builddep python3` will bring in git-core as well, because we use git to apply patches, but that is probably not explicit enough.)

Comment (hroncok) [reply .devcontainer/Dockerfile:12]: Invoking dnf with `--setopt=install_weak_deps=False` will also save space. Let me try sending a PR and see if it works.

Comment (hroncok) [reply .devcontainer/Dockerfile:12]: https://github.com/python/cpython/pull/103283
